### PR TITLE
Payload validation

### LINKED ISSUE #3266: GET request payload validation error
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure it has not already been reported

### Fastify version

3.20.1

### Plugin version

_No response_

### Node.js version

14.15

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

10.15.6

### Description

Fastify currently parses the request payload from `DELETE` requests but not from `GET` requests even though [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4.3) states the same for both methods.

> A payload within a DELETE request message has no defined semantics;
   sending a payload body on a DELETE request might cause some existing
   implementations to reject the request.

and

> A payload within a GET request message has no defined semantics;
   sending a payload body on a GET request might cause some existing
   implementations to reject the request.

The fact that Fastify does not parse the request payload of a `GET` request, but at the same time allows the developer to define a validation scheme for a `GET` endpoint, adds to confusing errors for the developer. So Fastify tries to validate the request payload, which was not parsed before. Thus, the client always receives the error `"body should be object"`since its request payload was not parsed and is, therefore `null` and the validation can only fail.

### Steps to Reproduce

```javascript
const app = require('fastify')()

app.get('/', {
    schema: {
        body: {
            name: {
                type: 'string'
            }
        }
    }
}, function (req, reply) {
   // req.body === null
   reply.send('ok')
})

app.listen(8080)

```

Now when a client makes the following request:

```http
GET / HTTP/1.1
Host: localhost:8080
Content-Type: application/json
Content-Length: 26

{
    "name": "testName"
}
```

it always receives the response:

```json
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": "body should be object"
}
```

### Expected Behavior

Fastify should therefore either parse the payload of `GET` requests, give developers a warning when they specify a body validation schema, or simply skip the body validation if it is a `GET` request.

### PR DESCRIPTION
**Pull request by matthyk**
Closes #3266. An error is now thrown if a body validation schema is specified for `GET` or `HEAD` routes. The docs also explain now that the request body validation is only available for `POST, PUT, PATCH, DELETE and OPTION` requests. The content-type parser documentation also now provides a more detailed explanation of when the request payload is parsed and when it is not.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/route.js [hunk 1]
```diff
@@ -137,11 +137,19 @@ function buildRouting (options) {
         if (supportedMethods.indexOf(method) === -1) {
           throw new Error(`${method} method is not supported!`)
         }
+
+        if ((method === 'GET' || method === 'HEAD') && opts.schema && opts.schema.body) {
```
Comment (climba03003) [inline lib/route.js:141]: I think we can extract this part of code as a function. It becomes large and duplicated.

**Review by mcollina — APPROVED** [review]
lgtm

File: test/schema-examples.test.js [hunk 2]
```diff
@@ -431,7 +431,7 @@ test('Example - schemas examples', t => {
     }
   }
 
-  fastify.get('/', {
+  fastify.post('/', {
```
Comment (Eomm) [inline test/schema-examples.test.js:434]: Could you update the validation-and-serialization.md file?

This test file contains all the snippet that are in that document

File: test/route.test.js [hunk 3]
```diff
@@ -1326,3 +1326,60 @@ test('Will not try to re-createprefixed HEAD route if it already exists and expo
...
+    fastify.route({
+      method: 'HEAD',
+      path: '/shouldThrow',
+      schema: {
+        body: {}
+      },
+      handler: function (req, reply) {
+        reply.send({ hello: 'world' })
+      }
+    })
```
Comment (Eomm) [inline test/route.test.js:1364]: Could you check the error message?

File: test/schema-examples.test.js (hunk 2 above)
Comment (matthyk) [reply test/schema-examples.test.js:434]: Yes, I will have a look at it tomorrow!

File: lib/route.js [hunk 4]
```diff
@@ -137,11 +141,19 @@ function buildRouting (options) {
         if (supportedMethods.indexOf(method) === -1) {
```
Comment (climba03003) [inline lib/route.js:141]: Can you include all the logic of validating `method` and `schema.body` inside the extracted function. It should be something similar to `validateBodyLimitOption`. You may name it as `validateMethodAndSchemaBodyOption`.

File: lib/route.js [hunk 5]
```diff
@@ -137,11 +141,19 @@ function buildRouting (options) {
...
           throw new Error(`${method} method is not supported!`)
         }
+
+        if (shouldNotContainBodyValidationSchema(method, opts.schema)) {
+          throw new Error('Body validation schemas for GET and HEAD routes are not supported!')
```
Comment (zekth) [inline lib/route.js:146]: Wouldn't it be good to throw the route name in the exception?

File: test/route.test.js [hunk 6]
```diff
@@ -1326,3 +1326,60 @@ test('Will not try to re-createprefixed HEAD route if it already exists and expo
...
+  }, new Error('Body validation schemas for GET and HEAD routes are not supported!'))
+})
+
+test('[HEAD, GET] route with body schema should throw', t => {
```
Comment (RafaelGSS) [inline test/route.test.js:1368]: Can you add a test using shorthand method? `route.get('/shouldThrow', { schema: { body: {} } }, handler)`?

File: test/schema-examples.test.js (hunk 2 above)
Comment (matthyk) [reply test/schema-examples.test.js:434]: I've looked through the docs but haven't found a snippet anywhere that I've changed that shows up in the docs. Can you show me what I should change?

**Review by climba03003 — APPROVED** [review]
LGTM.

**Review by mcollina — APPROVED** [review]
lgtm

**Review by zekth — APPROVED** [review]
Lgtm

File: test/schema-examples.test.js (hunk 2 above)
Comment (Eomm) [reply test/schema-examples.test.js:434]: This test was checking the schemas only
https://github.com/fastify/fastify/blob/next/docs/Validation-and-Serialization.md#examples

not the schema+routing

so you are right: no update needed

**Review by Eomm — APPROVED** [review]

**Pull request merged**